File: 09_gpu_map/problems/00_map_result/map_result.f90
program map_result
  use omp_lib
  implicit none
  real :: t, a(3)
  character(len=32) :: arg
  integer :: i

  if (command_argument_count() > 0) then
     call get_command_argument(1, arg)
     read(arg, *) t
  else
     t = 10.0
  end if
  a = (/ t + 1, t + 2, t + 3 /)
  ! TODO: GPUで更新した結果がホストに反映されるよう, target 構文に map(tofrom: ...) を付けよ.
  print "(a,f12.6)", "GPU: t = ", t
  print "(a,3f12.6)", "GPU: a = ", a(1), a(2), a(3)
  t = t * 2.0
  do i = 1, 3
     a(i) = a(i) * 2.0
  end do
  ! TODO: 上で始めた target 領域を閉じる (!$omp end target).
  print "(a,f12.6)", "CPU: t = ", t
  print "(a,3f12.6)", "CPU: a = ", a(1), a(2), a(3)
end program map_result

File: 09_gpu_map/problems/00_map_result/map_result.cpp
#include <cstdio>
#include <cstdlib>
#include <omp.h>

int main(int argc, char ** argv) {
  float t = (argc > 1 ? atof(argv[1]) : 10.0);
  float a[3] = { t + 1, t + 2, t + 3 };
  // TODO: GPUで更新した結果がホストに反映されるよう, target 構文に map(tofrom: ...) を付けよ.
  {
    printf("GPU: t = %f\n", t);
    printf("GPU: a = { %f, %f, %f }\n", a[0], a[1], a[2]);
    t *= 2.0;
    for (int i = 0; i < 3; i++) a[i] *= 2.0;
  }
  printf("CPU: t = %f\n", t);
  printf("CPU: a = { %f, %f, %f }\n", a[0], a[1], a[2]);
  return 0;
}
